# Global Terrorism Database (GTD) - Exploratory Data Analysis and Data Visualization

This project analyzes terrorism incidents across the world using the [Global Terrorism Database (GTD)](https://www.start.umd.edu/gtd/). The dataset includes detailed information on terrorist events from 1970 to mid-2021.

We'll be performing exploratory data analysis, cleaning inconsistencies, and exploring trends through visualization.

## 🛠️ Environment Setup & Loading Datasets

We are using 2 datasets, one is the main dataset that covers incidents from 1970 to 2020 and the other is an extension of that dataset that covers events that occurred between January and June 2021.

We'll be loading and concatenating these into a single DataFrame for analysis.

In [105]:
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt 
import seaborn as sns
import warnings

warnings.filterwarnings("ignore")

sns.set_theme(style= "whitegrid")
sns.set_palette("Set2")

#Ensures all plots are displayed in the notebook.
%matplotlib inline

#Setting the maximum number of rows and columns displayed on printing.
pd.set_option("display.max_rows",50)
pd.set_option("display.max_columns", 100)

In [64]:
terror_2020 = pd.read_csv(r"data\globalterrorismdb_0522dist.csv", index_col = "eventid")
terror_2021 = pd.read_csv(r"data\globalterrorismdb_2021Jan-June_1222dist.csv", index_col = "eventid")

#Checking shape of each dataframe.
print("GTD (till 2020):",terror_2020.shape)
print("GTD (2021 Jan-June):", terror_2021.shape)

terror = pd.concat([terror_2020,terror_2021])
print("GTD (combined):", terror.shape)

#To preview the dataset.
terror.head()

GTD (till 2020): (209706, 134)
GTD (2021 Jan-June): (4960, 134)
GTD (combined): (214666, 134)


,iyear,imonth,iday,approxdate,extended,resolution,country,country_txt,region,region_txt,provstate,city,latitude,longitude,specificity,vicinity,location,summary,crit1,crit2,crit3,doubtterr,alternative,alternative_txt,multiple,success,suicide,attacktype1,attacktype1_txt,attacktype2,attacktype2_txt,attacktype3,attacktype3_txt,targtype1,targtype1_txt,targsubtype1,targsubtype1_txt,corp1,target1,natlty1,natlty1_txt,targtype2,targtype2_txt,targsubtype2,targsubtype2_txt,corp2,target2,natlty2,natlty2_txt,targtype3,...,weaptype2,weaptype2_txt,weapsubtype2,weapsubtype2_txt,weaptype3,weaptype3_txt,weapsubtype3,weapsubtype3_txt,weaptype4,weaptype4_txt,weapsubtype4,weapsubtype4_txt,weapdetail,nkill,nkillus,nkillter,nwound,nwoundus,nwoundte,property,propextent,propextent_txt,propvalue,propcomment,ishostkid,nhostkid,nhostkidus,nhours,ndays,divert,kidhijcountry,ransom,ransomamt,ransomamtus,ransompaid,ransompaidus,ransomnote,hostkidoutcome,hostkidoutcome_txt,nreleased,addnotes,scite1,scite2,scite3,dbsource,INT_LOG,INT_IDEO,INT_MISC,INT_ANY,related
eventid,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
197000000001,1970,7,2,NaN,0,NaN,58,Dominican Republic,2,Central America & Caribbean,National,Santo Domingo,18.456792,-69.951164,1.0,0,NaN,NaN,1,1,1,0,NaN,NaN,0.0,1,0,1,Assassination,NaN,NaN,NaN,NaN,14,Private Citizens & Property,68.0,Named Civilian,NaN,Julio Guzman,58.0,Dominican Republic,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.0,NaN,NaN,0.0,NaN,NaN,0,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,PGIS,0,0,0,0,NaN
197000000002,1970,0,0,NaN,0,NaN,130,Mexico,1,North America,Federal,Mexico city,19.371887,-99.086624,1.0,0,NaN,NaN,1,1,1,0,NaN,NaN,0.0,1,0,6,Hostage Taking (Kidnapping),NaN,NaN,NaN,NaN,7,Government (Diplomatic),45.0,"Diplomatic Personnel (outside of embassy, cons...",Belgian Ambassador Daughter,"Nadine Chaval, daughter",21.0,Belgium,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,0.0,NaN,NaN,0,NaN,NaN,NaN,NaN,1.0,1.0,0.0,NaN,NaN,NaN,Mexico,1.0,800000.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,PGIS,0,1,1,1,NaN
197001000001,1970,1,0,NaN,0,NaN,160,Philippines,5,Southeast Asia,Tarlac,Unknown,15.478598,120.599741,4.0,0,NaN,NaN,1,1,1,0,NaN,NaN,0.0,1,0,1,Assassination,NaN,NaN,NaN,NaN,10,Journalists & Media,54.0,Radio Journalist/Staff/Facility,Voice of America,Employee,217.0,United States,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.0,NaN,NaN,0.0,NaN,NaN,0,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,PGIS,-9,-9,1,1,NaN
197001000002,1970,1,0,NaN,0,NaN,78,Greece,8,Western Europe,Attica,Athens,37.997490,23.762728,1.0,0,NaN,NaN,1,1,1,0,NaN,NaN,0.0,1,0,3,Bombing/Explosion,NaN,NaN,NaN,NaN,7,Government (Diplomatic),46.0,Embassy/Consulate,NaN,U.S. Embassy,217.0,United States,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Explosive,NaN,NaN,NaN,NaN,NaN,NaN,1,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,PGIS,-9,-9,1,1,NaN
197001000003,1970,1,0,NaN,0,NaN,101,Japan,4,East Asia,Fukouka,Fukouka,33.580412,130.396361,1.0,0,NaN,NaN,1,1,1,-9,NaN,NaN,0.0,1,0,7,Facility/Infrastructure Attack,NaN,NaN,NaN,NaN,7,Government (Diplomatic),46.0,Embassy/Consulate,NaN,U.S. Consulate,217.0,United States,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Incendiary,NaN,NaN,NaN,NaN,NaN,NaN,1,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,PGIS,-9,-9,1,1,NaN


## Dataset Overview

Lets begin by understanding the structure of the dataset.

- Each row is an instance of a terrorist event around the world.
- Each column represents one of the features of the event.

We'll be inspecting for duplicated rows, inconsistent data, and the kind of numeric data in different columns.

In [65]:
print(f"The dataset has {terror.shape[0]} rows and {terror.shape[1]} columns. \n") #To get the number of rows and columns

print("Dataset info:")
print(terror.info(),"\n")

print("Columns: ")
print(terror.columns.tolist(),"\n")

print("Number of duplicated rows:",terror.duplicated().sum() )

print("\nFirst 5 rows: ")
terror.head()

The dataset has 214666 rows and 134 columns. 

Dataset info:
<class 'pandas.core.frame.DataFrame'>
Index: 214666 entries, 197000000001 to 202107020037
Columns: 134 entries, iyear to related
dtypes: float64(54), int64(22), object(58)
memory usage: 221.1+ MB
None 

Columns: 
['iyear', 'imonth', 'iday', 'approxdate', 'extended', 'resolution', 'country', 'country_txt', 'region', 'region_txt', 'provstate', 'city', 'latitude', 'longitude', 'specificity', 'vicinity', 'location', 'summary', 'crit1', 'crit2', 'crit3', 'doubtterr', 'alternative', 'alternative_txt', 'multiple', 'success', 'suicide', 'attacktype1', 'attacktype1_txt', 'attacktype2', 'attacktype2_txt', 'attacktype3', 'attacktype3_txt', 'targtype1', 'targtype1_txt', 'targsubtype1', 'targsubtype1_txt', 'corp1', 'target1', 'natlty1', 'natlty1_txt', 'targtype2', 'targtype2_txt', 'targsubtype2', 'targsubtype2_txt', 'corp2', 'target2', 'natlty2', 'natlty2_txt', 'targtype3', 'targtype3_txt', 'targsubtype3', 'targsubtype3_txt', 'corp3', 'ta

,iyear,imonth,iday,approxdate,extended,resolution,country,country_txt,region,region_txt,provstate,city,latitude,longitude,specificity,vicinity,location,summary,crit1,crit2,crit3,doubtterr,alternative,alternative_txt,multiple,success,suicide,attacktype1,attacktype1_txt,attacktype2,attacktype2_txt,attacktype3,attacktype3_txt,targtype1,targtype1_txt,targsubtype1,targsubtype1_txt,corp1,target1,natlty1,natlty1_txt,targtype2,targtype2_txt,targsubtype2,targsubtype2_txt,corp2,target2,natlty2,natlty2_txt,targtype3,...,weaptype2,weaptype2_txt,weapsubtype2,weapsubtype2_txt,weaptype3,weaptype3_txt,weapsubtype3,weapsubtype3_txt,weaptype4,weaptype4_txt,weapsubtype4,weapsubtype4_txt,weapdetail,nkill,nkillus,nkillter,nwound,nwoundus,nwoundte,property,propextent,propextent_txt,propvalue,propcomment,ishostkid,nhostkid,nhostkidus,nhours,ndays,divert,kidhijcountry,ransom,ransomamt,ransomamtus,ransompaid,ransompaidus,ransomnote,hostkidoutcome,hostkidoutcome_txt,nreleased,addnotes,scite1,scite2,scite3,dbsource,INT_LOG,INT_IDEO,INT_MISC,INT_ANY,related
eventid,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
197000000001,1970,7,2,NaN,0,NaN,58,Dominican Republic,2,Central America & Caribbean,National,Santo Domingo,18.456792,-69.951164,1.0,0,NaN,NaN,1,1,1,0,NaN,NaN,0.0,1,0,1,Assassination,NaN,NaN,NaN,NaN,14,Private Citizens & Property,68.0,Named Civilian,NaN,Julio Guzman,58.0,Dominican Republic,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.0,NaN,NaN,0.0,NaN,NaN,0,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,PGIS,0,0,0,0,NaN
197000000002,1970,0,0,NaN,0,NaN,130,Mexico,1,North America,Federal,Mexico city,19.371887,-99.086624,1.0,0,NaN,NaN,1,1,1,0,NaN,NaN,0.0,1,0,6,Hostage Taking (Kidnapping),NaN,NaN,NaN,NaN,7,Government (Diplomatic),45.0,"Diplomatic Personnel (outside of embassy, cons...",Belgian Ambassador Daughter,"Nadine Chaval, daughter",21.0,Belgium,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,0.0,NaN,NaN,0,NaN,NaN,NaN,NaN,1.0,1.0,0.0,NaN,NaN,NaN,Mexico,1.0,800000.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,PGIS,0,1,1,1,NaN
197001000001,1970,1,0,NaN,0,NaN,160,Philippines,5,Southeast Asia,Tarlac,Unknown,15.478598,120.599741,4.0,0,NaN,NaN,1,1,1,0,NaN,NaN,0.0,1,0,1,Assassination,NaN,NaN,NaN,NaN,10,Journalists & Media,54.0,Radio Journalist/Staff/Facility,Voice of America,Employee,217.0,United States,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.0,NaN,NaN,0.0,NaN,NaN,0,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,PGIS,-9,-9,1,1,NaN
197001000002,1970,1,0,NaN,0,NaN,78,Greece,8,Western Europe,Attica,Athens,37.997490,23.762728,1.0,0,NaN,NaN,1,1,1,0,NaN,NaN,0.0,1,0,3,Bombing/Explosion,NaN,NaN,NaN,NaN,7,Government (Diplomatic),46.0,Embassy/Consulate,NaN,U.S. Embassy,217.0,United States,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Explosive,NaN,NaN,NaN,NaN,NaN,NaN,1,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,PGIS,-9,-9,1,1,NaN
197001000003,1970,1,0,NaN,0,NaN,101,Japan,4,East Asia,Fukouka,Fukouka,33.580412,130.396361,1.0,0,NaN,NaN,1,1,1,-9,NaN,NaN,0.0,1,0,7,Facility/Infrastructure Attack,NaN,NaN,NaN,NaN,7,Government (Diplomatic),46.0,Embassy/Consulate,NaN,U.S. Consulate,217.0,United States,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Incendiary,NaN,NaN,NaN,NaN,NaN,NaN,1,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,PGIS,-9,-9,1,1,NaN


## 🧹 Dataset Cleaning

From the overview we have understood that the GTD dataset is complex and large with 134 columns and 200,000+ entries, including a mix of textual, numerical and date-related data.

Before conducting any analysis we must ensure that the data is tidy, consistent, and relevant to our goal.


### Checking for missing values
Lets look for columns with a large number of missing values to decide what to drop or keep.

In [66]:
null_counts = terror.isnull().sum() # returns a series with each column and number of nulls
sorted_null_counts = terror.isnull().sum().sort_values(ascending=False)

total_null_count = null_counts.sum()
total_cell_count = terror.shape[0] * terror.shape[1]
total_null_percent = (total_null_count/total_cell_count)*100 # will give the percentage of the null cells in entire dataset.

print("The percentage of null entries in the GTD dataset is:", total_null_percent)

null_percent = (sorted_null_counts / len(terror)) * 100
missing_df = pd.DataFrame({'Missing Values': sorted_null_counts, 'Percent': null_percent})
missing_df[missing_df['Missing Values'] > 0]

columns_mtm50 = missing_df[missing_df.Percent >50].index
spared_columns = ['divert','ransomnote','ransompaidus','ransomamtus', 'ransompaid','ransomamt','resolution','compclaim', 'nhours', 'gsubname','ndays','nreleased','natlty2_txt', 'natlty2','related', 'alternative', 'alternative_txt','propvalue', 'motive', 'propcomment', 'location','propextent', 'propextent_txt', 'ransom']
columns_to_drop = [col for col  in columns_mtm50 if col not in spared_columns]
#dropping columns with more that 50% missing percent.
print(f"\nDropped {len(columns_to_drop)} columns.")
print("Dropped columns:", columns_to_drop)

#terror.drop(columns=columns_to_drop, inplace =True) 

The percentage of null entries in the GTD dataset is: 56.348067132682765

Dropped 53 columns.
Dropped columns: ['gsubname3', 'weapsubtype4', 'weapsubtype4_txt', 'weaptype4_txt', 'weaptype4', 'claimmode3', 'claimmode3_txt', 'gsubname2', 'claim3', 'guncertain3', 'gname3', 'attacktype3', 'attacktype3_txt', 'claimmode2', 'claimmode2_txt', 'corp3', 'targsubtype3_txt', 'targsubtype3', 'natlty3_txt', 'natlty3', 'target3', 'targtype3', 'targtype3_txt', 'weapsubtype3', 'weapsubtype3_txt', 'weaptype3', 'weaptype3_txt', 'claim2', 'guncertain2', 'gname2', 'kidhijcountry', 'attacktype2', 'attacktype2_txt', 'approxdate', 'corp2', 'hostkidoutcome_txt', 'hostkidoutcome', 'targsubtype2_txt', 'targsubtype2', 'weapsubtype2_txt', 'weapsubtype2', 'target2', 'targtype2_txt', 'targtype2', 'weaptype2', 'weaptype2_txt', 'nhostkidus', 'nhostkid', 'claimmode', 'claimmode_txt', 'addnotes', 'scite3', 'scite2']


We've dropped 53 columns that were selected on the basis of having more than 50% null peercent. Some columns that have necessary information were spared. (can be found in list spared_columns)

### Fixing Dates

In [82]:
#Lets display the date fields in the dataframe.

print(terror[["iyear","iday","imonth","approxdate"]])

KeyError: "None of [Index(['iyear', 'iday', 'imonth', 'approxdate'], dtype='object')] are in the [columns]"

Here we can see that the "approxdate" column is basically the iyear, iday and imonth in a date format.

So let's work on renaming the "approxdate" column to "eventdate" and parse the content from iyear, imonth and iday to the "eventdate" field as datetime format.

In [68]:
#Displaying the value counts for each unique entry in the iyear, imonth and iday columns.

print(terror.imonth.value_counts().sort_index())
print(terror.iday.value_counts().sort_index())
print(terror.iyear.value_counts().sort_index())

imonth
0        20
1     18062
2     16755
3     18285
4     18393
5     20664
6     18781
7     18783
8     18030
9     16471
10    17913
11    16916
12    15593
Name: count, dtype: int64
iday
0      891
1     7366
2     7014
3     7048
4     7214
5     6983
6     6862
7     7083
8     6872
9     7192
10    7162
11    7084
12    7144
13    7144
14    7088
15    7657
16    7197
17    6967
18    7017
19    7015
20    7084
21    6923
22    6840
23    6830
24    6829
25    6981
26    6820
27    7053
28    7140
29    6484
30    6004
31    3678
Name: count, dtype: int64
iyear
1970      651
1971      471
1972      567
1973      473
1974      581
        ...  
2017    11364
2018     9853
2019     8537
2020     8438
2021     4960
Name: count, Length: 51, dtype: int64


Here we can see that in both "iyear" and "imonth" columns there are 0 entries which likely denote the first day of the month, so well focus on changing those entries to 1 for seamless date parsing.

In [69]:
terror["imonth"] = terror["imonth"].replace(0,1)
terror["iday"] = terror["iday"].replace(0,1)

Now we can work on parsing the dates to the "approxdate" column and rename it to "eventdate"

In [72]:
#renaming approxdate to eventdate and iyear, imonth and iday to match pandas requirement for parsing
terror.rename(columns={"approxdate":"eventdate", "iyear":"year","imonth":"month","iday":"day"}, inplace=True)

terror["eventdate"] =pd.to_datetime(terror[["year","month","day"]], errors="coerce") #when giving dataframe columns as input, pandas automatically assumes the fields are year, month and date.

#checking how many dates are null
terror.eventdate.isna().sum()

np.int64(0)

### Changing DataTypes

Some columns in the GTD Dataset might need datatype conversions to work with the data, lets find out what those columns are and how we can edit them.

In [97]:
# To view all columns with respective datatypes
print(terror.dtypes)

#We are parsing the resolution column to datetime format.
terror.resolution.unique()
terror["resolution"] = pd.to_datetime(terror["resolution"], format= "%d/%m/%Y", errors="coerce")

year                           int64
month                          int64
day                            int64
eventdate             datetime64[ns]
extended                       int64
resolution            datetime64[ns]
country                        int64
country_txt                   object
region                         int64
region_txt                    object
provstate                     object
city                          object
latitude                     float64
longitude                    float64
specificity                  float64
vicinity                       int64
location                      object
summary                       object
crit1                          int64
crit2                          int64
crit3                          int64
doubtterr                      int64
alternative                  float64
alternative_txt               object
multiple                     float64
success                        int64
suicide                        int64
a

We've gone through the datatypes of all columns and changed resolution to the datetime datatype.

### Drop Duplicates

We will be checking the entries in the dataset and dropping duplicate rows based on eventid.

In [108]:
print("Number of duplicate entries:", terror.duplicated().sum())
terror = terror.drop_duplicates()

Number of duplicate entries: 9068


### In the Data Cleaning section we have:

- Identified columns with **high percentages of missing values (>50%)** and dropped them to reduce noise.
- Carefully **excluded important NLP-related columns** such as `ransomnote` and `ransompaid` from removal.
- Inspected data types for all columns and planned for **datatype conversions** where necessary to optimize memory and analysis.
- Parsed the `iyear`, `imonth`, and `iday` columns to create a new `approxdate` column, then **renamed it to `eventdate`**, to handle missing or invalid dates.
- Checked for and removed **fully duplicated rows** (~9,000 duplicates), ensuring each event is unique.


## Phase 3: Univariate Analysis